Separate staging/publish and unstaging/unpublish logics for block

mkimuram opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR changes staging/publish and unstaging/unpublish logics for block to be called from separate methods. Also, publish path is changed to per pod path from per volume path.

[Before]
- ```stageVolumeForBlock```: called from ```SetupDevice```
- ```publishVolumeForBlock```: called from ```SetupDevice```
- ```unstageVolumeForBlock```: called from ```TearDownDevice```
- ```unpublishVolumeForBlock```: called from ```TearDownDevice```
- publish path: ```plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}```

[After]
- ```stageVolumeForBlock```: called from ```SetupDevice```
- ```publishVolumeForBlock```: called from ```MapDevice```
- ```unstageVolumeForBlock```: called from ```TearDownDevice```
- ```unpublishVolumeForBlock```: called from ```UnmapDevice```
- publish path:  ```plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}/{podUid}```

Summary of discussion why this change needed is commented [here](https://github.com/kubernetes/kubernetes/issues/73773#issuecomment-462895942)

**Which issue(s) this PR fixes**:
Fixes #73773

**Special notes for your reviewer**:
/sig storage
@bswartz @msau42  @wongma7 @vladimirvivien 

**Does this PR introduce a user-facing change?**:
```release-note
All nodes need to be drained before upgrading Kubernetes cluster, because paths used for block volumes are changed in this release, so on-line upgrade of nodes aren't allowed. 
```

mkimuram commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -883,14 +889,18 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
-			return volumeToMount.GenerateError("MapVolume.EvalHostSymlinks failed", err)
+		// For in-tree drivers, actual device is attached either in WaitForAttach or SetUpDevice,
+		// not in MapDevice below, so devicePath could be checked here.
+		if !isCSI {
```
@msau42 @wongma7 

Do you have any better ideas to handle this, so that operation_generator code doesn't need to care whether it is CSI?
We might make ```SetUpDevice``` for CSI return dummy device paht and make all driver resolve symlink from ```volumeMapPath``` below, but I'm not sure if it is better than current one.

wongma7 replied:
for csi, is the devicePath not the staging path returned by SetUpDevice (NodeStage)?

wongma7 commented:
/ok-to-test

wongma7 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1096,11 +1116,18 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
-		// Execute tear down device
-		unmapErr := blockVolumeUnmapper.TearDownDevice(globalMapPath, deviceToDetach.DevicePath)
+		// Execute unmap device
+		unmapErr := blockVolumeUnmapper.UnmapDevice(globalMapPath, deviceToDetach.DevicePath)
```
should we call UnmapDevice in the per-pod UnmapVolume instead? Then what if we had all the in-tree plugins called og.blkUtil.UnmapDevice in their plugin.UnmapDevice, and CSI calls Unpublish in its plugin.UnmapDevice

mkimuram replied:
Thank you for pointing it out.
This will be a big change, so I wanted to make an agreement before go ahead.

I will try implementing this way and update the commit.

wongma7 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -248,22 +248,58 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
+	defer cancel()
+
 	// Call NodePublishVolume
+	stagingPath := m.getStagingPath()
```
should this be getPublishPath? currently getPublishPath is unused

mkimuram replied:
Staging path is used [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_block.go#L190) for publish call, it is due to [CSI spec](https://github.com/container-storage-interface/spec/blame/master/spec.md#L2030).

It's confusing, so I agree that we should change ```publishVolumeForBlock``` not to take ```stagingPath``` as an argument, and get it inside the function, as it is done for [```publishPath```](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_block.go#L172).

mkimuram reviewed (commented):
@wongma7 

Thank you for your comment.
I will update the commit as suggested.

mkimuram commented:
@wongma7 

Updated as suggested. PTAL

Note that ref-counting and releasing descriptor lock codes are also moved to CSI's ```UnmapDevice```, almost as they are. We might need to consider more to properly handle all cases including multiple pod mapping case.
(To get feed back early, I'm pushing this commit, after I only confirmed that this worked well with one pod mapping case.)

mkimuram commented:
/reopen

mkimuram commented:
/reopen

msau42 commented:
/remove-lifecycle rotten
/assign @jsafrane

gnufied commented:
/assign

mkimuram commented:
/retest

gnufied commented on pkg/volume/csi/csi_block.go:
```diff
@@ -252,22 +252,66 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
+	return "", nil
+}
+
+func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, volumeMapName string, podUID types.UID) error {
```
We are ignoring supplied `devicePath` parameter in this function and we only use return value of `NodePublish` as real device path. This may or may not be true and is a departure from previous version of this code and I am not sure if it is backward compatible change.  In old version of code - if `NodePublish` returned empty string then value returned by `WaitForAttach` was considered as canonical device path.

mkimuram replied:
@gnufied 

Thank you for your coment.

Fixed by taking `devicePath` returned by `SetupDevice`.
(There's no acutal device at `devicePath` by the time that `SetupDevice` returns, however it would be safe.)

wongma7 replied:
I don't follow, we don't use devicePath so should we set it to _? A short comment could explain that the plugin doesn't accept a devicePath because it cannot know the devicePath, i.e. whatever path on the node where the device actually exists, until after it has called NodePublish.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1070,6 +1080,20 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 		}
 
+		// For CSI driver, devicePath should be checked here by checking symlink from
+		// {volumeMapPath}/{volName}.
```
Why do we need to do this? Can we have `MapDevice` to return device path? For in-tree plugins it will simply return whatever was passed and for CSI it will return device path after publish.

mkimuram replied:
Fixed by directly using the `publishPath` as `devicePath`. (With ignoring error for resolving symlink.)
I thought `publishPath` also need to be resolved for containerized kubelet, so I put this logic here.
However, this assumption won't be true. (I think that `publishPath` could be handled by containerized kubelet.)

Only downside by doing this will be failure in handling error for resolving symlink for in-tree driver.

gnufied commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1070,6 +1080,20 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
+			if err != nil {
+				return volumeToMount.GenerateError("MapVolume.EvalHostSymlinks failed", err)
+			}
+		}
```
Also we need to fix when ASOW gets updated after node is staged etc. Currently both device is marked as mounted and volume is marked as mounted at the end. This needs to be split, how filesystem mounting is split.

mkimuram replied:
Fixe by moving  `MarkDeviceAsMounted` before `MapDevice`.

gnufied reviewed (commented):
I just reviewed mapdevice flow. I am yet to take a look at Unmap flow.

gnufied commented on pkg/volume/volume.go:
```diff
@@ -181,6 +181,9 @@ type BlockVolumeUnmapper interface {
 	TearDownDevice(mapPath string, devicePath string) error
+
+	// UnmapDevice unmaps the block device path
+	UnmapDevice(globalUnmapPath, podDeviceUnmapPath, volName string, podUID types.UID) error
```
I am thinking that we should call this function `UnmapPodDevice` or something like that - so as it is evident that this function is actually going to remove device/volume for the pod and not from global mount path.

mkimuram replied:
Then, do we need to also change `MapDevice` to `MapPodDevice`?

If everyone agrees, I will go ahead and change them at once.

wongma7 replied:
Agree to rename. It's quite confusing because there is the UnmapDevice operation, UnmapDevice util function, and here the UnmapDevice method for BlockVolume plugins.

mkimuram replied:
Renamed. Thank you for your explanation on the reason for rename.

mkimuram commented:
/retest

tedyu commented on pkg/volume/csi/csi_block.go:
```diff
@@ -252,21 +252,70 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
+	//TODO (vladimirvivien) implement better AccessModes mapping between k8s and CSI
+	accessMode := v1.ReadWriteOnce
+	if m.spec.PersistentVolume.Spec.AccessModes != nil {
+		accessMode = m.spec.PersistentVolume.Spec.AccessModes[0]
```
Do we care about other elements in AccessModes (len(m.spec.PersistentVolume.Spec.AccessModes) > 1) ?

mkimuram replied:
I agree that this should be imporved. 

But I think that this needs to be updated for both filesystem and block, in the same way.
(https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_mounter.go#L176)

So, this would be outside the scope of this PR.
Let's improve this in a separate PR.

tedyu commented on pkg/volume/csi/csi_block.go:
```diff
@@ -252,21 +252,70 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
+	defer cancel()
+
+	csiClient, err := m.csiClientGetter.Get()
```
It seems this can be lifted ahead of context.WithTimeout call (line 300).

mkimuram replied:
Thank you for your review and sorry for late response.

I confirmted `SetUpAt` and `TearDownAt` is calling `csiClientGetter.Get` before calling `WithTimeout`.
So, fixed as suggested.

tedyu commented on pkg/volume/csi/csi_block.go:
```diff
@@ -252,21 +252,70 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
-}
+	// Check that publishPath and devicePath matches
+	if publishPath != devicePath {
+		return errors.New(log("blockMapper.MapDevice failed because publishPath %q and devicePath %q doesn't match", publishPath, devicePath))
```
doesn't match -> don't match

mkimuram replied:
Fixed.

tedyu commented on pkg/volume/csi/csi_block.go:
```diff
@@ -331,31 +374,75 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
+	defer cancel()
+
+	csiClient, err := m.csiClientGetter.Get()
```
This can be lifted above line 429

mkimuram replied:
Fixed.

wongma7 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1054,6 +1054,12 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 
 		}
+		// For CSI driver, WaitForAttach returns attachment name as devicePath,
+		// not actual devicePath.
```
can you also add a little note here to say that for CSI the devicePath is returned by the below SetUpDevice instead.

wongma7 replied:
NVM, see comment below about the "// A plugin doesn't have attacher" comment. If we can clarify that comment then we don't need a duplicate comment above.

wongma7 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1054,6 +1054,12 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
+		if strings.HasPrefix(devicePath, "csi-") {
+			isCSI = true
+		}
 		// A plugin doesn't have attacher also needs to map device to global map path with SetUpDevice()
```
I know this is outside your PR but since we are here re-evaluating the meaning of devicePath etc. as it gets passed through various functions, its meaning dependent on CSI vs non-CSI, can we reword or remove this comment. TBH I don't know what it means

mkimuram replied:
Improved the comments.

wongma7 reviewed (commented):
will do more thorough review soon, thanks and sorry for not looking at it for months!!

wongma7 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -67,10 +68,10 @@ func (m *csiBlockMapper) getStagingPath() string {
+// Example: plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}/{podUid}
 func (m *csiBlockMapper) getPublishPath() string {
 	sanitizedSpecVolID := utilstrings.EscapeQualifiedName(m.specName)
-	return filepath.Join(m.plugin.host.GetVolumeDevicePluginDir(CSIPluginName), "publish", sanitizedSpecVolID)
```
shouldn't sanitizedSpecVolID be m.volumeID, not come from the PV name?

wongma7 replied:
the spec says publish/stage should be idempotent per volumeID & m.volumeID is taken from PV volumeHandle

mkimuram replied:
Changed to use volumeID.

mkimuram reviewed (commented):
@wongma7 

Thank you for your review. PTAL

wongma7 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -331,31 +373,96 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
-	// Call NodeUnstageVolume
-	stagingPath := m.getStagingPath()
-	if _, err := os.Stat(stagingPath); err != nil {
+	dataDir := getVolumeDeviceDataDir(m.specName, m.plugin.host)
```
this change was just reverted :/ https://github.com/kubernetes/kubernetes/pull/83379

mkimuram replied:
Thank you for quickly letting me know it!

wongma7 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -331,31 +373,96 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
 		}
 	} else {
-		err := m.unstageVolumeForBlock(ctx, csiClient, stagingPath)
+		err := m.unpublishVolumeForBlock(ctx, csiClient, publishPath)
```
since we have changed publishPath and stagingPath, there might be a backwards-compatibility issue.

Previously, we created a symlink in the pod's volume directory to `plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}`.
Now, we create a symlink in the pod's volume directory to `plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}/{podUID}`.

After an upgrade, we ask the plugin to unpublish `plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}/{podUID}` even if it, before the upgrade, published `plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}`. So `plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}` will be stuck in published state.

We may need to follow the symlink in the pod's volume directory to see if it links to the new path or the old path and call unpublish appropriately.

wongma7 replied:
proper upgrade procedure you are supposed to drain the node first. So I'm not sure this is worth fixing.

mkimuram replied:
Good investigation!
Then, should we add a caution like below in release notes?

```
Before this release, staging path and publish path for CSI block volume were as follows:
- Staging path: plugins/kubernetes.io/csi/volumeDevices/staging/{pvname}
- Publish path: plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}
In this release these are changed as follows:
- Staging path: plugins/kubernetes.io/csi/volumeDevices/staging/{volumeID}
- Publish path: plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}/{podUID}

As a result, upgrading Kubernetes cluster while any pods with CSI block volumes running on a node will cause these path inconsistent, which will lead failure in unstaging/unpublishing volume. 
Therefore, all nodes need to be drained before upgrading Kubernetes cluster.
```

Any improvements in wording are welcome.

wongma7 replied:
Thanks, IMO it doesn't need to be too detailed about the specific paths, the last two sentences reminding people that nodes need to be drained else unpublish/unstage will fail should be enough.

wongma7 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -331,31 +372,75 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
+		return err
+	}
+	if len(refs) > 0 {
+		return fmt.Errorf("blockMapper.UnmapPodDevice The device %q is still referenced from other Pods %v", globalMapPath, refs)
```
this seems wrong, I thought we want to call NodeUnpublish every time we call UnmapDevice from the UnmapVolume operation.

mkimuram replied:
This is intended to avoid releasing descriptor lock by calling `DetachFileDevice`, if other pods are still using this device.

For CSI plugins that don't implement node stage/unstage, actual detach of the device from the node will happen in node unpublish. So, this kind of plugins requires descriptor lock to be released before calling node unpublish. 
(Releasing descriptor lock before calling `TearDownDevice`, or node unstage, in `GenerateUnmapDeviceFunc` is too late for such plugins. Actually, CSI rbd puglin doesn't work well with this code.)

And this code seems to be incorrect for CSI plugins that implement node stage/unstage, as you pointed out.

I started to think that we might need to move node unpublish to `TearDownDevice` if CSI plugins don't implement node stage/unstage. WDYT?

wongma7 replied:
If a CSI plugin is capable of publishing the same volume ID to two different paths (MULTI_NODE) it must also be capable of unpublishing the same volume ID from them independently. The current code makes it so that if I NodePublish(vol-a, path-1) and NodePublish(vol-a, path-2), only one of NodeUnpublish(vol-a, path-1) and NodeUnpublish(vol-a, path-2) will ever be called, but I would expect both to be called eventually, and that's why I say something seems wrong.

I understand this logic is here because the file descriptor lock will prevent, for certain plugins, a NodeUnpublish(vol-a, path-x) from succeeding. I would rather see us change the file descriptor lock logic than move the unpublish call, because I think the way you have aligned CSI calls to operations makes the most sense still.

What if we put this code in operation generator's UnmapVolume (to before UnmapPodDevice is called) and removed it from operation generator's UnmapDevice?

mkimuram replied:
Thank you for your suggestion. I've changed as suggested. 
By this change, we don't need any special handlings in CSI codes for it.

Also, as we changed publish path, or `devicePath`,  to be per pod for CSI, we will be able to expect only one symlink to the `devicePath` for CSI driver in `unmapVolumeFunc`.
(On the other hand, in-tree driver would stack here if the same block volume is attached to multiple pods on the same node, until all the pods on the same node will be deleted (Unless devicePath become per pod, too.))

PTAL

mkimuram replied:
Ummm... With this code, number of symlinks won't be decremented because `UnmapPodDevice` isn't called, when more than one symlinks exist, as a result `UnmapPodDevice` will never be called for the above in-tree situation.

We might be able to move `UnmapBlockVolume` to `unmapVolumeFunc`  from driver's `UnmapPodDevice` to work around this, but I'm not sure this will be the right direction.

wongma7 replied:
> We might be able to move UnmapBlockVolume to unmapVolumeFunc from driver's UnmapPodDevice to work around this, but I'm not sure this will be the right direction.

I think something like this is the correct fix. Right now, it is implicitly required that every plugin must execute this block of code to put a symlink in globalMapPath: 
https://github.com/kubernetes/kubernetes/blob/8635f7a0b09301227470ca7c229fc1aab5458984/pkg/volume/util/util.go#L521
and them remove it later:
https://github.com/kubernetes/kubernetes/pull/74026/files#diff-c3acf7a7fec401d0a3cb4998088ee1e5R547

Instead of having this implicit requirement, and every plugin doing the same thing, we should move this logic outside of the plugin and into the operation_generator. The plugin implementation should not have to worry about the file lock or the globalMapPath, if we keep that logic isolated to operation_generator I think it will be much easier to understand.

mkimuram replied:
Thank you for your comment. I will go ahead and fix further.
Also, I found that descriptor lock is taken every time after map, so we should release it every time before map. (And, it could be inside map/unmap utility function.)

wongma7 replied:
> Also, as we changed publish path, or devicePath, to be per pod for CSI, we will be able to expect only one symlink to the devicePath for CSI driver in unmapVolumeFunc.

Hmm I am having trouble following, which symlink do you mean? globalMapPath is still per-volume, not per-pod https://github.com/kubernetes/kubernetes/pull/74026/files#diff-ed824558531f8d1f616bc2f976266b64R53 and publishPath used to be per-volume, now it is per-pod. 

globalMapPath is per-volume and is for ref-counting via symlinks to publishPath/devicePath
publishPath is per-pod and is for 
- symlinking to pod mount path (e.g. /var/lib/kubelet/pods/{podUID}/volumes/{escapeQualifiedPluginName}/{outerVolumeSpecName}/) 
- locking (AttachFileDevice)

We lock the publishPath after CSI NodePublish.
a. if the publishPath is a symlink to another file, that file will be locked
b. if the publishPath is its own file, it will be locked

Now, let's say the same volume is NodePublished again to a different path.
a. if the publishPath is a symlink to another file, that file is already locked: no problem, idempotent
b. if the publishPath is its own file, it will be locked.

What do we do when one of the volume needs to be unmapped.
a. if the publishPath is a symlink to another file, we should not unlock it
b. if the publishPath is its own file, we should unlock it

How do we distinguish between case a and case b...

wongma7 replied:
To summarize issue, the in-tree plugins could always safely assume that there is a 1:1 mapping between  volumeID and devicePath. But now there could be 1:many mapping between volumeID and devicePath.

wongma7 replied:
also there is a case c where publishPath is a bind mount to a /dev/a. in which case I don't think there needs to be a lock on /dev/a at all because device will be busy and can't be detached.

I am not an expert on why the lock is here in the first place.

mkimuram replied:
I was only thinking about UnmapPodDevice call for CSI plugin is not blocked by current reference count logic. 

As you explained, we need to think about adding another logics:
  - To block calling node unpublish for CSI plugins that don't have unstage if the device is still in use,
  - To block calling node unstage for CSI plugins that have unstage if the device is still in use.

I've manually tested how loopback device works with bind mount and symlink. Please see below:

<details>

```
[Bind mount case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}
# touch /tmp/publish/pod1
# touch /tmp/publish/pod2

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# mount --bind /dev/loop0 /tmp/publish/pod1
# mount --bind /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# ln -s /tmp/publish/pod1  /tmp/globalmap/pod1
# ln -s /tmp/publish/pod2  /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
brw-rw----. 1 root disk 7, 0 Oct  3 20:50 pod1
brw-rw----. 1 root disk 7, 0 Oct  3 20:50 pod2

# ls -l /tmp/globalmap/
total 0
lrwxrwxrwx. 1 root root 17 Oct  3 20:52 pod1 -> /tmp/publish/pod1
lrwxrwxrwx. 1 root root 17 Oct  3 20:53 pod2 -> /tmp/publish/pod2

6. Take descriptor lock
# losetup /dev/loop1 /tmp/publish/pod1
# losetup /dev/loop2 /tmp/publish/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/publish/pod1
/dev/loop2         0      0         0  0 /tmp/publish/pod2

7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/publish/pod1
/dev/loop2         0      0         0  0 /tmp/publish/pod2

# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /tmp/publish/pod2

# umount /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root    0 Oct  3 20:50 pod1
brw-rw----. 1 root disk 7, 0 Oct  3 21:14 pod2

# losetup -d /dev/loop2
# losetup 

# umount /tmp/publish/pod2
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root 0 Oct  3 20:50 pod1
-rw-r--r--. 1 root root 0 Oct  3 20:50 pod2

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/


[Symlink case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# ln -s /dev/loop0 /tmp/publish/pod1 
# ln -s /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# ln -s /tmp/publish/pod1  /tmp/globalmap/pod1
# ln -s /tmp/publish/pod2  /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  3 21:20 pod1 -> /dev/loop0
lrwxrwxrwx. 1 root root 10 Oct  3 21:20 pod2 -> /dev/loop0

# ls -l /tmp/globalmap/
total 0
lrwxrwxrwx. 1 root root 17 Oct  3 21:20 pod1 -> /tmp/publish/pod1
lrwxrwxrwx. 1 root root 17 Oct  3 21:20 pod2 -> /tmp/publish/pod2

6. Take descriptor lock
# losetup /dev/loop1 /tmp/publish/pod1
# losetup /dev/loop2 /tmp/publish/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /dev/loop0

# rm -f /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  3 21:20 pod2 -> /dev/loop0


# losetup -d /dev/loop2
# losetup 

# rm -f /tmp/publish/pod2
# ls -l /tmp/publish/
total 0

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/
```
</details>

mkimuram replied:
I'm thinking about making globalMap a bind mount instead of symlink and take descriptor lock for globalMap instead of devicePath (Then, it will be per pod lock, as well as ref count. Also, it will be idempotent).

If descriptor lock is taken to devicePath and CSI driver makes device available on publishPath as a symlink, we can't distinguish for which pod descriptor lock was taken, because losetup -j show the resolved path (Please see [Symlink case]'s step 6 in details sections of previous comment).

If we change globalMap to bind mount and take descriptor lock to it, we can see which descriptor lock is taken for which pod and who takes descriptor lock for the device (Please compare step 4 and step 6 in details sections of this comment and previous one).

[Current implementation]
````
                                    devicePath                            globalMap
                                   -------------                   ----------------------
1. actual device -- symlink    --> publishPath     <-- symlink --   glolbalMapPath/PodUID
2. actual device -- bind mount --> publishPath     <-- symlink --   glolbalMapPath/PodUID
3.                                 actual device   <-- symlink --   glolbalMapPath/PodUID
                                     ^--- descriptor lock
````

[This idea]
````
                                    devicePath                               globalMap
                                   -------------                       ----------------------
1. actual device -- symlink    --> publishPath     <-- bind mount --   glolbalMapPath/PodUID
2. actual device -- bind mount --> publishPath     <-- bind mount --   glolbalMapPath/PodUID
3.                                 actual device   <-- bind mount --   glolbalMapPath/PodUID
                                                                        ^--- descriptor lock
````

<details>

````
[Bind mount case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}
# touch /tmp/publish/pod1
# touch /tmp/publish/pod2

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# mount --bind /dev/loop0 /tmp/publish/pod1
# mount --bind /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# touch /tmp/globalmap/{pod1,pod2}
# mount --bind /tmp/publish/pod1 /tmp/globalmap/pod1
# mount --bind /tmp/publish/pod2 /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod2

# ls -l /tmp/globalmap/
total 0
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:33 pod2

# findmnt | grep pod
├─/tmp/publish/pod1                   devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
├─/tmp/publish/pod2                   devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
├─/tmp/globalmap/pod1                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
└─/tmp/globalmap/pod2                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755

6. Take descriptor lock
# losetup /dev/loop1 /tmp/globalmap/pod1
# losetup /dev/loop2 /tmp/globalmap/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2

# losetup -j /tmp/globalmap/pod1
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)

# losetup -j /tmp/globalmap/pod2
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)

7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2


# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2

# umount /tmp/globalmap/pod1
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root    0 Oct  4 12:34 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:40 pod2

# umount /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root    0 Oct  4 12:33 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:40 pod2

# losetup -d /dev/loop2
# losetup 

# umount /tmp/globalmap/pod2
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root 0 Oct  4 12:34 pod1
-rw-r--r--. 1 root root 0 Oct  4 12:34 pod2

# umount /tmp/publish/pod2
# ls -l /tmp/publish/
total 0
-rw-r--r--. 1 root root 0 Oct  4 12:33 pod1
-rw-r--r--. 1 root root 0 Oct  4 12:33 pod2

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/


[Symlink case]
1. Prepare files to be used
# dd if=/dev/zero of=/tmp/file1 bs=512 count=1
# mkdir /tmp/{publish,globalmap}

2. Prepare pseudo device file
# losetup /dev/loop0 /tmp/file1

3. Prepare pseudo published device
# ln -s /dev/loop0 /tmp/publish/pod1 
# ln -s /dev/loop0 /tmp/publish/pod2

4. Prepare pseudo globalmap file
# ln -s /tmp/publish/pod1  /tmp/globalmap/pod1
# ln -s /tmp/publish/pod2  /tmp/globalmap/pod2

# touch /tmp/globalmap/{pod1,pod2}
# mount --bind /tmp/publish/pod1 /tmp/globalmap/pod1
# mount --bind /tmp/publish/pod2 /tmp/globalmap/pod2

5. Check current status (Symlinks under globalmap path reference to different publish path.)
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1

# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  4 12:44 pod1 -> /dev/loop0
lrwxrwxrwx. 1 root root 10 Oct  4 12:44 pod2 -> /dev/loop0

# ls -l /tmp/globalmap/
total 0
brw-rw----. 1 root disk 7, 0 Oct  4 12:44 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:44 pod2

# findmnt | grep pod
├─/tmp/globalmap/pod1                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755
└─/tmp/globalmap/pod2                 devtmpfs[/loop0] devtmpfs   rw,nosuid,seclabel,size=484564k,nr_inodes=121141,mode=755


6. Take descriptor lock
# losetup /dev/loop1 /tmp/publish/pod1
# losetup /dev/loop2 /tmp/publish/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

# losetup /dev/loop1 /tmp/globalmap/pod1
# losetup /dev/loop2 /tmp/globalmap/pod2

# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2

# losetup -j /tmp/globalmap/pod1
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)

# losetup -j /tmp/globalmap/pod2
/dev/loop1: [0005]:9426815 (/tmp/globalmap/pod1)
/dev/loop2: [0005]:9426815 (/tmp/globalmap/pod2)


7. Check that device can't be deleted before descriptor lock is released
# losetup -d /dev/loop0
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop1         0      0         0  0 /tmp/globalmap/pod1
/dev/loop2         0      0         0  0 /tmp/globalmap/pod2


# losetup -d /dev/loop1
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         1  0 /tmp/file1
/dev/loop2         0      0         0  0 /dev/loop0

# umount /tmp/globalmap/pod1
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root    0 Oct  4 12:45 pod1
brw-rw----. 1 root disk 7, 0 Oct  4 12:48 pod2

# rm -f /tmp/publish/pod1
# ls -l /tmp/publish/
total 0
lrwxrwxrwx. 1 root root 10 Oct  4 12:44 pod2 -> /dev/loop0

# losetup -d /dev/loop2
# losetup 

# umount /tmp/globalmap/pod2
# ls -l /tmp/globalmap
total 0
-rw-r--r--. 1 root root 0 Oct  4 12:45 pod1
-rw-r--r--. 1 root root 0 Oct  4 12:45 pod2

# rm -f /tmp/publish/pod2
# ls -l /tmp/publish/
total 0

8. Clean up
# rm -rf /tmp/file1 /tmp/publish/ /tmp/globalmap/
````
</details>

mkimuram replied:
@wongma7 

Changed as commented above (Please also see the details below, for how pathes look like when block volumes are mapped). I will work on further debugging and improvement, but I would appreciate it if you could also review the current code.

<details>

```
Events:
  Type     Reason                  Age        From                     Message
  ----     ------                  ----       ----                     -------
  Normal   Scheduled               <unknown>  default-scheduler        Successfully assigned default/pod001-block to 127.0.0.1
  Normal   SuccessfulAttachVolume  25s        attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278"
  Warning  FailedMapVolume         16s        kubelet, 127.0.0.1       MapVolume.MapBlockVolume failed for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278" : device not found
  Normal   SuccessfulMountVolume   15s        kubelet, 127.0.0.1       MapVolume.MapPodDevice succeeded for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278" globalMapPath "/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev"
  Normal   SuccessfulMountVolume   15s        kubelet, 127.0.0.1       MapVolume.MapPodDevice succeeded for volume "pvc-82051086-eb98-4dd0-941a-1512bc438278" volumeMapPath "/var/lib/kubelet/pods/c34639f6-0a8f-458f-bf8e-24a284788187/volumeDevices/kubernetes.io~csi"
  Normal   Pulled                  14s        kubelet, 127.0.0.1       Container image "k8s.gcr.io/busybox:1.24" already present on machine
  Normal   Created                 13s        kubelet, 127.0.0.1       Created container container001
  Normal   Started                 12s        kubelet, 127.0.0.1       Started container container001


# ls -l /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev
total 0
brwxrwxrwx. 1 root disk 251, 80 Oct  5 00:37 c34639f6-0a8f-458f-bf8e-24a284788187

# ls -l /var/lib/kubelet/pods/c34639f6-0a8f-458f-bf8e-24a284788187/volumeDevices/kubernetes.io~csi
total 4
lrwxrwxrwx. 1 root root 175 Oct  5 00:37 pvc-82051086-eb98-4dd0-941a-1512bc438278 -> /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/0001-0009-rook-ceph-0000000000000001-3745bc30-e708-11e9-b798-0242ac110005/c34639f6-0a8f-458f-bf8e-24a284788187

# losetup | grep '/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev'
/dev/loop7         0      0         0  0 /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187                                        0

# findmnt /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187
TARGET                                                                                                                                     SOURCE          FSTYPE   OPTIONS
/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187 devtmpfs[/rbd5] devtmpfs rw,seclabel,size=16460544

# lsblk | grep rbd5
rbd5   251:80   0    1G  0 disk 

# mount | grep /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187
devtmpfs on /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-82051086-eb98-4dd0-941a-1512bc438278/dev/c34639f6-0a8f-458f-bf8e-24a284788187 type devtmpfs (rw,seclabel,size=16460544k,nr_inodes=4115136,mode=755)
```

</details>

tedyu replied:
I am curious about the FailedMapVolume event shown above.
Did you find out the cause of it ?

Thanks

mkimuram replied:
Thank you for pointing it out.

I've found that this is due to the difference of losetup command's output depending on the arguments. I've fixed it in the latest commit. PTAL

(In the previous code, `AttachFileDevice` failed because `makeLoopDevice` failed to find the device just created.)

wongma7 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1316,31 +1316,18 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
-		klog.V(4).Infof("UnmapDevice: deviceToDetach.DevicePath: %v", deviceToDetach.DevicePath)
-		loopPath, err := og.blkUtil.GetLoopDevice(deviceToDetach.DevicePath)
+		// This should be done before calling UnmapDevice, because some plugins that do local detach
+		// in UnmapDownDevice will fail in detaching device due to the refcnt on the loopback device.
```
should it be TearDownDevice instead of UnmapDownDevice/UnmapDevice

wongma7 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1316,31 +1316,18 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
-		loopPath, err := og.blkUtil.GetLoopDevice(deviceToDetach.DevicePath)
+		// This should be done before calling UnmapDevice, because some plugins that do local detach
+		// in UnmapDownDevice will fail in detaching device due to the refcnt on the loopback device.
+		err = og.blkUtil.DetachFileDevice(deviceToDetach.DevicePath)
```
so for CSI, this call will do nothing, is that right. Since CSI did AttachFileDevice for each of its publish paths, rather than at the deviceToDetach.DevicePath...

wongma7 replied:
NVM, we discussed more in detail above the purpose of DetachFileDevice.

wongma7 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1217,6 +1222,31 @@ func (og *operationGenerator) GenerateUnmapVolumeFunc(
+		}
+		if len(refs) > 1 {
+			err = fmt.Errorf("The device %q is still referenced from other Pods %v", globalUnmapPath, refs)
+			return volumeToUnmount.GenerateError("UnmapVolume failed", err)
```
we could DetachFileDevice only if len(refs) == 1 but still proceed to UnmapPodDevice if len(refs) > 1?

wongma7 replied:
No, that won't work. Continuing discussion in original thread...

mkimuram commented:
> Consider logging the err before it is overwritten (on next line).

Added a log for it. (Actually, this is a normal path, so I out dented this code to make it clearer.)

wongma7 commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -93,14 +125,35 @@ func (v VolumePathHandler) RemoveLoopDevice(device string) error {
 }
 
-func parseLosetupOutputForDevice(output []byte) (string, error) {
+func parseLosetupOutputForDevice(output []byte, path string, exactMatch bool) (string, error) {
```
I wouldn't mind 2 functions. Maybe parseLosetupAssociatedOutput... (-j) and parseLosetupFileOutput... (-f)

mkimuram replied:
Thank you for suggestion.

I was thinking about providing a way to call `DetachFileDevice` that can release descriptor lock regardless of how its backend path looks like (For example, to be used before calling `TearDownDevice` in `operationGenerator` to ensure no descriptor lock remains). However, we would rather avoid such a situation that needs force releasing locks (To do so, we might need to add a check if all descriptor lock is released before calling `UnmapDevice` so that there will be no descriptor lock).  So, let's change `DetachFileDevice` to release the path that exact matches, then we will no longer need the `exactMatch` flag, which is much simpler.

Also, I found that we might need to add to `MapPodDevice` and `UnmapPodDevice` implementations to all in-tree plugins that support block volume so that their device paths to be per pod (I'm thinking about making all in-tree plugins also bind mount their actual devices to per pod paths in `MapPodDevice`). Without them, descriptor lock won't work well for in-tree plugins for multi attaching the same volume case.

wongma7 replied:
Do we need the file descriptor lock at all if we are already bind mounting the path /dev/asdf to globalMapPath? It is getting complicated again with the in-tree vs csi case :(

mkimuram replied:
The last commit will make in-tree driver also make use of per pod descriptor lock.
As you indicate, this is to just make descriptor lock to be per pod (ref count is already per pod), so if we remove descriptor lock logic, we can remove most of these codes.  

(I haven't tested enough, but just bind mounting device doesn't seem to prevent device from being detached, as loop back device do.)

@msau42 

Do you think that we should remove descriptor lock for simplicity? Or any good ideas to improve?

wongma7 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -59,17 +58,15 @@ func (m *csiBlockMapper) GetGlobalMapPath(spec *volume.Spec) (string, error) {
 }
```
globalMapPath should be based on volumeID too, not pv name.

mkimuram replied:
@wongma7 

I tried to fix both `getVolumeDevicePluginDir` and `getVolumeDeviceDataDir` to be based on `volumeID`. However, in `NewBlockVolumeUnmapper`, `volumeID` is gotten by calling `loadVolumeData` with specifying `unmapper.specName`, or volName. So, at least, `getVolumeDeviceDataDir` won't be able to be based on `volumeID`.

Should we make only `getVolumeDevicePluginDir` based on `volumeID`, or keep the paths as they are?
I think that pv name and volumeID is 1:1, so this change won't be a mandatory. Then, we would be easier to debug the volume if we keep device and volume data under the same directory.

WDYT?

wongma7 replied:
Why is pv name and volumeID 1:1? Can't PV name be anything? I assume that whoever wrote this code had a good reason to think pvname == volumeID but I am not so sure. I thought volumeID is supposed to be 1:1 with volumeHandle, not PV name?

wongma7 replied:
It seems like variables are poorly named. "specVolID" is PV name. https://github.com/kubernetes/kubernetes/blob/39724859b524e345881eb8ef458f0d5a2f25bd9a/pkg/volume/csi/csi_mounter.go#L45 . This is not the same thing as `volume_id` in the CSI spec, that is volumeHandle?!

wongma7 replied:
The issue I see is it's possible (but unlikely) for the same volume_id to be used by two different PVs, in which case there will be two globalMapPaths for the same `volume_id` meaning NodeStage may be called early.

I think you can keep the paths as they are to make this PR simpler, and you can leave the sorting out of paths as a follow-up for me...

mkimuram replied:
@wongma7 

It seems that we have been using [`spec.Name()`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L472) as an ID for volume in [GetPodDeviceMapPath](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gcepd/gce_pd_block.go#L178) in in-tree plugins. So, we should be able to use `m.specName` as CSI plugin's ID for volume.

The value of `m.specName` is like `pvc-82051086-eb98-4dd0-941a-1512bc438278`, so we won't need to sanitize it by calling `utilstrings.EscapeQualifiedName`.
Also, we should change the comment from `{pvname}` to `{specName}` to avoid confusion.

If you agree. I will change as so.

wongma7 replied:
I agree 👍

wongma7 replied:
While I am not enough of a linux pro to understand why it works, I tried the fd lock thing today and it works as described by the comment before AttachFileDevice. I will need to check if it still works for the new bind mounted case...

1. Start some nonsense container that periodically writes to the device
docker run --name test --device /dev/nvme2n3 nginx bash -c 'i=0; while true; do echo $i; ((i++)); echo $i > ./hello.txt; dd if=./hello.txt of=/dev/nvme2n3 bs=64 count=1; sleep 5; done'; docker rm test
2. Detach the device, then it will get:
dd: failed to open '/dev/nvme2n3': No such device or address
3. Attach a different device, it will resume writing to it. This is the problem the comment for the lock described.

If you setup a loop device to keep the fd open:

1. docker run --name test --device /dev/nvme2n3 nginx bash -c 'i=0; while true; do echo $i; ((i++)); echo $i > ./hello.txt; dd if=./hello.txt of=/dev/nvme2n3 bs=64 count=1; sleep 5; done'; docker rm test
2. sudo losetup -f --show /dev/nvme2n3
3. Detach the device, then it will get:
dd: failed to open '/dev/nvme2n3': No such device or address
4. Attach a different device, it will continue to get the above error.

mkimuram commented:
/retest

mkimuram commented:
@wongma7 

I've fixed reconstruction logic and now all tests pass. PTAL
I will squash the commit once, after you checked the changes.

We might need to further discuss:
- Whether we need descriptor lock at all (I hope that current implementation is still simple enough, though),
- Whether we should (and could) also make podDeviceMapPath a bind mount, instead of symlink
- Further refactoring, like renaming variables, paths and methods, in this PR.

wongma7 commented on pkg/volume/util/util.go:
```diff
@@ -500,28 +500,85 @@ func MakeAbsolutePath(goos, path string) string {
+				globalUnmapPath, string(podUID), err)
+		}
+	} else if len(lockPath) != 0 {
+		return fmt.Errorf("Descriptor lock for globalUnmapPath:%s, podUID: %s is still taken. Wait for another retry to release it",
```
How could this be possible if we just did DetachFileDevice in the lines above? What will be different in the next retry that it may succeed?

mkimuram replied:
I think that `MapBlockVolume` could be called more than once, so there would be a case that multiple descriptor locks are taken to the same `lockPath`, just for one volume-pod mapping. (I haven't actually hit this case, though.)

If above assumption is true, we might also need to return error if bindmount still exist after `UnmapDevice` for `globalUnmapPath` below this code, because bindmount could be created multiple times (On the other hand, symlink can't be created multiple times, so we won't need the same check for `podDeviceUnmapPath`).

wongma7 replied:
If the path is the same, i.e. a loop device already exists whose backing file is equal to the path, AttachFileDevice should no-op: https://github.com/kubernetes/kubernetes/blob/a3ccea9d8743f2ff82e41b6c2af6dc2c41dc7b10/pkg/volume/util/volumepathhandler/volume_path_handler_linux.go#L34. So we don't need to worry about that case.

Similarly, if the path is the same and you try to bind mount twice, wouldn't it error in MapBlockVolume?
mount --bind a b # 1st time
mount --bind a b # 2nd time, error?

mkimuram replied:
Thank you for your comment.

For bind mount, I've added a code to check if the path has already been bind mounted in `mapBindMountDevice`. So, it would be as safe as loop back device. 
However, IMHO, we don't take a lock between check and creation, so there could be a very small chance for creation logic to be called multiple times. 

And in such a situation, both `losetup` and `mount --bind` allow creating them every time, as seen below:

<details>

```
[losetup]
# losetup -f --show /dev/loop0 
/dev/loop1
# losetup -f --show /dev/loop0 
/dev/loop2
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

[bind mount]
# touch /tmp/bindmount
# mount --bind /dev/loop0 /tmp/bindmount 
# mount --bind /dev/loop0 /tmp/bindmount 
# mount | grep /tmp/bindmount
devtmpfs on /tmp/bindmount type devtmpfs (rw,nosuid,seclabel,size=485080k,nr_inodes=121270,mode=755)
devtmpfs on /tmp/bindmount type devtmpfs (rw,nosuid,seclabel,size=485080k,nr_inodes=121270,mode=755)
```

</details>

wongma7 replied:
Thanks for testing. operation_generator/executor should insulate us from multiple creations  happening concurrently but it doesn't hurt to have these checks.

mkimuram replied:
Good to hear that operation_generator/executor avoid such a concurrent call. Then, I agree to delete this unnecessary code.

wongma7 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -252,22 +247,66 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
 
-	// Call NodePublishVolume
-	publishPath, err := m.publishVolumeForBlock(ctx, csiClient, accessMode, csiSource, attachment, stagingPath)
+	return m.getPublishPath(), nil
```
This is a little bit confusing. We NodeStage the volume to staging path but return publish path, which is currently empty, with the expectation that MapPodDevice will later NodePublish the volume there.

wongma7 replied:
TBH I think it's fine, I don't want to refactor this again, but would appreciate some comments for future readers.

mkimuram replied:
Thank you for pointing it out. I will add comments, here.

wongma7 commented on pkg/volume/csi/csi_block_test.go:
```diff
@@ -306,56 +330,41 @@ func TestBlockMapperMapDevice(t *testing.T) {
+		t.Fatalf("mapper failed to close dummy file in publishPath: %v", err)
 	}
 
 	// Map device to global and pod device map path
```
this test comment no longer accurate

wongma7 commented:
>  Whether we need descriptor lock at all (I hope that current implementation is still simple enough, though),

Yes, let's keep it. It works as intended from my testing. And it still works even when you create the lock on a bind mount of the original file instead of the original file itself e.g. /dev/nvme2n1. I.e. if I detach and re-attach a device it won't go to /dev/nvme2n1. If I DON'T create the lock, it does replace /dev/nvme2n1.

> Whether we should (and could) also make podDeviceMapPath a bind mount, instead of symlink

I recall some debates about using one or the other but don't know the details. So I am fine with keeping it as-is. Note we did change globalMapPath from symlink to bind mount.

> Further refactoring, like renaming variables, paths and methods, in this PR.
I think we are OK on this front. Left some comments. Overall, the code is much less confusing than before!

wongma7 commented:
/lgtm

Will let you do some squashing as you see fit.

I think we've achieved the goals from https://github.com/kubernetes/kubernetes/issues/73773#issuecomment-462895942

> Let's move the publish call back to MapDevice so that it is called per pod and not per volume. This is to be consistent with how NodePublish/Stage calls work for filesystem volumes. This may also require adding UnmapDevice to the BlockVolume interface

Done. Now Stage is in SetUpDevice and Publish is in MapPodDevice.

> We may need to clarify in the CSI spec the semantics around multi access on a single node.

We may still want to do this. But it doesn't block this PR. With per-pod publish/staging paths, it's up to the plugin to decide if it can handle "MULTI_NODE*" or not.

> We are unable to simplify the number of paths for now. globalMapPath and podPath are needed for Kubernetes to refcnt the number of pods using the device to gate the unstage call. The path returned by nodeStageVolume is not actually a device, it's a directory where a plugin keeps information about the device.
> It's a little odd that the Kubernetes refcnting mechanism is exposed into the plugin. Maybe that can be cleaned up later.

Done, we've pulled globalMapPath and podDeviceMapPath from outside the plugin code. We are effectively reversing https://github.com/kubernetes/kubernetes/pull/64094/files. CSI has staging and publish paths, kubernetes has globalMapPath and podDeviceMapPath. Responsibilities are clear. CSI plugin puts things in staging and publish paths. kubernetes puts things in globalMapPath (add/remove files as pods are added/removed) and podDeviceMapPath (expose CSI publish paths on it).

mkimuram reviewed (commented):
@wongma7 

Thank you for your review and testing!
I will squash the commit and add some comments to the codes.

Also, please give me feedback on `UnmapBlockVolume` logic.

msau42 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -31,7 +31,6 @@ import (
-	ioutil "k8s.io/kubernetes/pkg/volume/util"
 	utilstrings "k8s.io/utils/strings"
 )
 
```
Can you add some comments at the top here that summarize the new high level sequence?

msau42 reviewed (commented):
Thank you for working on these changes!

I want to make sure we test a real csi driver with these changes (not just hostpath). @davidz627 can you update the PD csi driver specs and enable block tests so we can test it with this change?

/assign @saad-ali

davidz627 commented:
> @davidz627 can you update the PD csi driver specs and enable block tests so we can test it with this change?

https://github.com/kubernetes/kubernetes/pull/84312

saad-ali commented:
/assign

mkimuram commented:
@wongma7

I've added comments, fixed unit tests, and remove the redundant codes. PTAL

For unit test fix, let me explain a bit:
  - `MapPodDevice` no longer requires publishPath to be created before test, because it doesn't create symlink for the path,
  - Tests for `TearDownDevice` and `UnmapPodDevice` seemed to have tested without stagingPath and publishPath, so no NodeUnstage/NodeUnpblish wasn't called. Tests passed because initial state, or lack of the key, is regarded as called. (To test them properly, we should add the key before calling these methods and confirm that key doesn't exist after these methods are called).

mkimuram commented:
Rebased with the latest master to reflect #84312 change. Thanks @davidz627 for updating CSI GCE pd driver.

mkimuram commented:
/retest

mkimuram commented:
@wongma7 @msau42 @saad-ali @davidz627 

All tests pass now including tests for newly added csi gce pd. PTAL

bswartz commented:
I have tested this with my CSI driver and didn't detect any breakages.

mkimuram commented:
@bswarz 

Thank you for your testing and reporting the result!

davidz627 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -14,6 +14,55 @@ See the License for the specific language governing permissions and
+
+Summary of block volume related CSI driver's methods are as follows:
+ - GetGlobalMapPath returns a global map path,
+ - GetPodDeviceMapPath returns a pod device map path and filename,
```
can you clarify what a `global map path` and `pod device map path and filename` are, or what they are used for

davidz627 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -14,6 +14,55 @@ See the License for the specific language governing permissions and
+   MountVolume
+          +--------> GetGlobalMapPath
+          +--------> SetupDevice ------------------------------------------------> NodeStageVolume
+          +-----------------------------------------------> MarkDeviceAsMounted
```
do we need this many dashes? It makes it hard to read on smaller screens

davidz627 replied:
in general for this and the below "After successful MountVolume..." section I think it's better to not rely on formatting based on spacing as it can be difficult to read and may differ depending on dev configuration. Looks like this could probably just be simplified into a nested bulleted list type structure with the occasional arrow

davidz627 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -51,34 +99,31 @@ type csiBlockMapper struct {
 
 // getStagingPath returns a staging path for a directory (on the node) that should be used on NodeStageVolume/NodeUnstageVolume
-// Example: plugins/kubernetes.io/csi/volumeDevices/staging/{pvname}
+// Example: plugins/kubernetes.io/csi/volumeDevices/staging/{specName}
```
what is `specName`

davidz627 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -252,22 +296,71 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
-	publishPath, err := m.publishVolumeForBlock(ctx, csiClient, accessMode, csiSource, attachment, stagingPath)
+	// In SetUpDevice, NodeStageVolume is called, but NodePublishVolume isn't,
+	// therefore, there will be no device available in publishPath, yet.
+	// However, GenerateMapVolumeFunc expects SetUpDevice to return the path that the device
```
this comment feels off - if the interface is expecting something that we can't provide, that feels a lot like we're putting logic in the wrong place.

davidz627 commented on pkg/volume/csi/csi_block.go:
```diff
@@ -252,22 +296,71 @@ func (m *csiBlockMapper) SetUpDevice() (string, error) {
+func (m *csiBlockMapper) MapPodDevice() error {
+	// Get csiSource from spec
+	if m.spec == nil {
+		klog.Error(log("blockMapper.MapPodDevice spec is nil"))
```
comment applies throughout all code - please don't `klog` then return an `error` especially with different contexts. We should *only* return an error with all relevant context and let the top level callers log errors

davidz627 commented on pkg/volume/fc/fc.go:
```diff
@@ -416,8 +416,8 @@ func (b *fcDiskMapper) SetUpDevice() (string, error) {
-func (b *fcDiskMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, volumeMapName string, podUID types.UID) error {
-	return util.MapBlockVolume(devicePath, globalMapPath, volumeMapPath, volumeMapName, podUID)
+func (b *fcDiskMapper) MapPodDevice() error {
+	return nil
```
looks like most (many?) plugins have `nil` implementations for this. Should this be a separate interface that these plugins just don't implement? Something like `BlockVolumeMapper`

davidz627 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -1079,26 +1087,13 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
+		resolvedDevicePath, err := hu.EvalHostSymlinks(devicePath)
 		if err != nil {
-			return volumeToMount.GenerateError("MapVolume.AttachFileDevice failed", err)
+			// For CSI plugin, device won't be available at devicePath until MapPodDevice is called,
```
this may ignore real errors - could you instead do a check like `if !csi then evalhostsymlinks else do nothing`?

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler.go:
```diff
@@ -98,25 +103,105 @@ func (v VolumePathHandler) MapDevice(devicePath string, mapPath string, linkName
+	}
+
+	// Unmount file
+	mounter := &mount.SafeFormatAndMount{Interface: mount.New(""), Exec: mount.NewOSExec()}
```
why are we making a new mounter here, seems fishy. Shouldn't we be using a mounter passed in from elsewhere that may have specific interface or exec? How will you unit test this

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler.go:
```diff
@@ -163,70 +247,42 @@ func (v VolumePathHandler) IsSymlinkExist(mapPath string) (bool, error) {
+	fi, err := os.Lstat(mapPath)
+	if err == nil {
+		// If file exits and it's device, return true and no error
+		if fi.Mode()&os.ModeDevice == os.ModeDevice {
```
is there no helper function to do this?

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler.go:
```diff
@@ -163,70 +247,42 @@ func (v VolumePathHandler) IsSymlinkExist(mapPath string) (bool, error) {
+// On other cases, return false with error from Lstat().
+func (v VolumePathHandler) IsBindMountExist(mapPath string) (bool, error) {
+	fi, err := os.Lstat(mapPath)
+	if err == nil {
```
can we flip this logic for ease of reading?

```
if err != nil{
  error stuff
}
do other stuff
```

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -47,6 +50,27 @@ func (v VolumePathHandler) AttachFileDevice(path string) (string, error) {
+	loopPath, err := v.GetLoopDevice(path)
+	if err != nil {
+		if err.Error() == ErrDeviceNotFound {
+			klog.Warningf("DetachFileDevice: Couldn't find loopback device which takes file descriptor lock. device path: %q", path)
```
logic in this func could be cleaned up a bit. For example I think this logs a warning and goes to the end of the function retuning `nil`. Is it supposed to just do nothing? It's unclear from the warning that this is ok (maybe it should say - "assuming file device already cleaned up" or osmething)

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -47,6 +50,27 @@ func (v VolumePathHandler) AttachFileDevice(path string) (string, error) {
+		if len(loopPath) != 0 {
+			err = removeLoopDevice(loopPath)
+			if err != nil {
+				return err
```
add context for `err` when available

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -93,16 +124,97 @@ func (v VolumePathHandler) RemoveLoopDevice(device string) error {
-	// /dev/loop1: [0073]:148662 (/dev/sda)
-	device := strings.TrimSpace(strings.SplitN(string(output), ":", 2)[0])
+	// losetup -j {path} returns device in the format:
+	// /dev/loop1: [0073]:148662 (/dev/{globalMapPath}/{pod1-UUID})
```
what is `globalMapPath`, isn't it different per volume/plugin? But in this case it's not even a parameter to func so its not clear what this is referring to

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -93,16 +124,97 @@ func (v VolumePathHandler) RemoveLoopDevice(device string) error {
+	}
+	s = matched
+
+	device := strings.TrimSpace(strings.SplitN(s, ":", 2)[0])
```
lot of magic numbers in this function. why are we splitting specifically with a max of 2 substrings but only taking the 0th? What is the 0th?

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -93,16 +124,97 @@ func (v VolumePathHandler) RemoveLoopDevice(device string) error {
+// FindGlobalMapPathUUIDFromPod finds {pod uuid} bind mount under globalMapPath
+// corresponding to map path symlink, and then return global map path with pod uuid.
+// ex. mapPath symlink: pods/{podUid}}/{DefaultKubeletVolumeDevicesDirName}/{escapeQualifiedPluginName}/{volumeName} -> /dev/sdX
+//     globalMapPath/{pod uuid} bind mount: plugins/kubernetes.io/{PluginName}/{DefaultKubeletVolumeDevicesDirName}/{volumePluginDependentPath}/{pod uuid} -> /dev/sdX
```
these `{}` variables for example `{DefaultKubeletVolumeDevicesDirName}` are not discoverable in this scope - what are they referring to?

davidz627 commented on pkg/volume/util/volumepathhandler/volume_path_handler_linux.go:
```diff
@@ -93,16 +124,97 @@ func (v VolumePathHandler) RemoveLoopDevice(device string) error {
+	return false, nil
+}
+
+func getDeviceMajorMinor(path string) (string, error) {
```
sidenote: I thought this PR was just splitting up `staging/publish` and `unstage/unpublish` for block - why am I seeing so much new functionality all over the place? It's not clear to mewhat this stuff is doing

davidz627 reviewed (commented):
just a quick surface level 1st pass review. Overall comment is that this probably could benefit from being seperated into some smaller PRs. It's quite bulky right now and sometimes its not clear why certain changes are being made

mkimuram commented:
@davidz627 

Thank you for your comment. I will re-organize the commits and explain the intention per commit.

mkimuram committed:
Change publish path for CSI block volume per pod

This change is to allow CSI driver to publish the same volume for multipe pods on the same node.

mkimuram committed:
Separate staging/publish and unstaging/unpublish logics for block

mkimuram commented:
@davidz627 

I've created #84625 to separate the commits and explain the intention of the code to each commit comment. This is still work in progress, because no logic change from this PR's version, so I will continue working on to change the code per your comments.

I hope that above PR at least answer to your two comments below:

> Overall comment is that this probably could benefit from being seperated into some smaller PRs.

> why am I seeing so much new functionality all over the place? It's not clear to mewhat this stuff is doing


I would appreciate it if you could suggest me how we separate it to small PRs. (Should it be per commit in #84625?)

davidz627 commented:
I don't really understand 100% what's happening enough to have suggestions on how to split this up into smaller PRs. It may be that the answer is it can't and some commit separation is helpful enough for review :) I'm just concerned since theres a lot of code moving around and it's not clear which parts are refactoring and which are net-new. The PR title says logic is being "seperated" but I see some completely new functions with significant logic. It might be worth seperating out the refactor into one PR where things just move around, and another PR for each of the issues that need to be fixed

mkimuram commented:
@davidz627 

Thank you for your comment. 

I think that the last three commits in #84625 is actually separating the logic as described in the PR tittle, and other commits are prerequisite for it. So, let's split them into two, and split them more if needed.

mkimuram commented:
Rebased again on top of #84747 after #84660 is merged. Waiting for #84747 to be merged.

@saad-ali 

Can this also be targeted to 1.17?  (Although, @jsafrane 's time zone is already midnight.) 

The last discussion for this PR is https://github.com/kubernetes/kubernetes/pull/84625#discussion_r345346651 . I'm pushing the latest version here, so that we can compare this PR and #84625 . I believe that this PR should be ready to be merged, if others agree.)

Any comments? @davidz627 @msau42 @jingxu97

mkimuram commented:
Now #84747 is merged. I will rebase this PR with the latest master.

mkimuram commented:
@saad-ali 

Rebased. Now it should be ready to be merged. PTAL

saad-ali reviewed (approved):
/lgtm
/approve
/milestone v1.17

saad-ali commented:
/priority important-soon
